sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingQueryManager.scala

```diff
@@ -206,21 +219,21 @@ class StreamingQueryManager private[sql] (sparkSession: SparkSession) {
       recoverFromCheckpointLocation: Boolean = true,
       trigger: Trigger = ProcessingTime(0),
       triggerClock: Clock = new SystemClock()): StreamingQuery = {
-    activeQueriesLock.synchronized {
-      val name = userSpecifiedName match {
-        case Some(n) =>
-          if (activeQueries.values.exists(_.name == userSpecifiedName.get)) {
-            throw new IllegalArgumentException(
-              s"Cannot start query with name $n as a query with that name is already active")
-          }
-          n
-        case None => null
+    userSpecifiedName foreach { name =>
+      activeQueriesLock.synchronized {
+        if (activeQueries.values.exists(_.name == name) || pendingQueryNames.contains(name)) {
+          throw new IllegalArgumentException(
+            s"Cannot start query with name $name as a query with that name is already active")
+        }
+        pendingQueryNames += name
       }
+    }
+    try {
       val checkpointLocation = userSpecifiedCheckpointLocation.map { userSpecified =>
         new Path(userSpecified).toUri.toString
       }.orElse {
         df.sparkSession.sessionState.conf.checkpointLocation.map { location =>
-          new Path(location, name).toUri.toString
+          new Path(location, userSpecifiedName.getOrElse(UUID.randomUUID().toString)).toUri.toString
```

Comment:
Can you add a unit test to verify this behavior. The test can something like 
- with userspecified location, see if it creates the right directory
- without userspecified location and without SQL conf, see if it creates a unique CP directory
- without userspecified location and with SQL conf, see if it creates a unique CP directory 
This can be in the StreamingQueryManager and would be easier if creating the query becomes a separate function.